### PR TITLE
Expose minizip API to allow creating zips using scripts

### LINKED ISSUE #1069: Expose support for packing/unpacking ZIP archives to the scripting API
**Describe the project you are working on:**
Node based image and material editor.

**Describe the problem or limitation you are having in your project:**
I would like to be able to pack and unpack user configuration to a single zip file. I will for sure find another usage for it later.

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
I will be able to do what I cannot do now.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
Inspiration can be taken from gdunzip (link below). Zip functionality can be a method in File class (which could zip files and directories).

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It cannot be done with a few lines of script.

**Is there a reason why this should be core and not an add-on in the asset library?:**
The plugin in GDScript was made but it's only unzip, it doesn't do CRC checks and it's not very fast. The engine already uses some C library (still minizip?) so it shouldn't be hard to expose it to GDScript to get native performance.

Related links [3701](https://github.com/godotengine/godot/issues/3701), [22440](https://github.com/godotengine/godot/issues/22440) and [gdunzip](https://github.com/jellehermsen/gdunzip).

*Bugsquad edit (keywords for easier searching): extract, create*

### PR DESCRIPTION
**Pull request by MisterMX**
Refactored and rebased version of #34444.

Closes https://github.com/godotengine/godot-proposals/issues/1069

CC @mhilbrunner

### CODE REVIEW

**Comment by fire** [comment]
Gordon requested this feature, and I realized that we had an old salvageable pull request from years ago.

**Comment by fire** [comment]
Documentation stubs must be generated, and then documentation can be optionally written.

File: modules/zip/zip_packer.cpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	zipfi.tmz_date.tm_mday = time.day;
+	zipfi.tmz_date.tm_min = time.minute;
+	zipfi.tmz_date.tm_mon = time.month - 1;
+	zipfi.tmz_date.tm_sec = time.second;
+	zipfi.tmz_date.tm_year = time.year;
+	zipfi.dosDate = 0;
+	zipfi.external_fa = 0;
+	zipfi.internal_fa = 0;
+
+	int ret = zipOpenNewFileInZip(zf, path.utf8().get_data(), &zipfi, NULL, 0, NULL, 0, NULL, Z_DEFLATED, Z_DEFAULT_COMPRESSION);
```
Comment (touilleMan) [inline modules/zip/zip_packer.cpp:68]: Don't we need to check if zf is not null, and if that's the case to call `zipCloseFileInZip` before doing the new open ?

Comment (MisterMX) [reply modules/zip/zip_packer.cpp:69]: Thanks for the hint. I added an additional null check.

**Review by touilleMan — APPROVED** [review]
lgtm 🎉

**Comment by fire** [comment]
I can’t review this now, but I support merging this into main

File: modules/zip/zip_reader.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+protected:
+	static void _bind_methods();
+
+public:
+	Error open(String path);
+	Error close();
+
+	PackedStringArray get_files();
+	PackedByteArray read_file(String path, bool case_sensitive);
```
Comment (akien-mga) [inline modules/zip/zip_reader.h:54]: Parameters should start with `p_`:
```suggestion
	Error open(String p_path);
	Error close();

	PackedStringArray get_files();
	PackedByteArray read_file(String p_path, bool p_case_sensitive);
```
(Same in other classes and in the .cpp)

File: modules/zip/zip_reader.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+/*************************************************************************/
+
+#ifndef ZIP_READER_H
+#define ZIP_READER_H
+
+#include "core/object/ref_counted.h"
+
+#include "core/io/file_access.h"
+#include "core/os/os.h"
+#include "thirdparty/minizip/unzip.h"
```
Comment (akien-mga) [inline modules/zip/zip_reader.h:38]:
```suggestion
#include "core/io/file_access.h"
#include "core/object/ref_counted.h"

#include "thirdparty/minizip/unzip.h"
```

File: modules/zip/zip_reader.h (hunk 2 above)
Comment (akien-mga) [reply modules/zip/zip_reader.h:54]: That's a convention only for C++ code though, the names given in `bind_method` should not have the `p_` prefix, so they're fine as is.

File: modules/zip/zip_packer.h [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+/*************************************************************************/
+
+#ifndef ZIP_PACKER_H
+#define ZIP_PACKER_H
+
+#include "core/object/ref_counted.h"
+
+#include "core/io/file_access.h"
+#include "core/os/os.h"
+#include "thirdparty/minizip/zip.h"
```
Comment (akien-mga) [inline modules/zip/zip_packer.h:38]:
```suggestion
#include "core/object/ref_counted.h"
#include "core/io/file_access.h"

#include "thirdparty/minizip/zip.h"
```

File: modules/zip/zip_packer.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "core/io/zip_io.h"
+
+#include "zip_packer.h"
```
Comment (akien-mga) [inline modules/zip/zip_packer.cpp:33]:
```suggestion
#include "zip_packer.h"

#include "core/io/zip_io.h"
```

File: modules/zip/register_types.h [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "modules/register_module_types.h"
+
+void initialize_zip_module(ModuleInitializationLevel p_level);
+void uninitialize_zip_module(ModuleInitializationLevel p_level);
```
Comment (akien-mga) [inline modules/zip/register_types.h:34]:
```suggestion
#ifndef ZIP_REGISTER_TYPES_H
#define ZIP_REGISTER_TYPES_H

#include "modules/register_module_types.h"

void initialize_zip_module(ModuleInitializationLevel p_level);
void uninitialize_zip_module(ModuleInitializationLevel p_level);

#endif // ZIP_REGISTER_TYPES_H
```

File: modules/zip/register_types.cpp [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#include "register_types.h"
+#include "core/object/class_db.h"
+#include "zip_packer.h"
+#include "zip_reader.h"
```
Comment (akien-mga) [inline modules/zip/register_types.cpp:34]:
```suggestion
#include "register_types.h"

#include "core/object/class_db.h"
#include "zip_packer.h"
#include "zip_reader.h"
```

**Comment by akien-mga** [comment]
Looks pretty good to me, left some style review comments.

The documentation should likely be filled too so that users know how to use the new APIs.

One thing I'm not sure of is whether there's any merit in making this a module, or if it should directly be in `core/io/`. It relies on the same dependencies as `core/io` and it's not a lot of code so I don't think the modularity brings much here. WDYT @godotengine/core?

**Comment by MisterMX** [comment]
@akien-mga, I included your suggestions. Let me know if you want this to be in core.

File: modules/zip/zip_reader.cpp [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+	PackedStringArray arr;
+	arr.resize(s.size());
+	int idx = 0;
+	for (const List<String>::Element *E = s.front(); E; E = E->next()) {
+		arr.set(idx++, E->get());
+	}
+	return arr;
+}
+
+PackedByteArray ZIPReader::read_file(String path, bool case_sensitive) {
```
Comment (akien-mga) [inline modules/zip/zip_reader.cpp:87]: Should use `p_` prefix like in the header (same above).

File: modules/zip/zip_packer.h [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+		APPEND_CREATE = 0,
+		APPEND_CREATEAFTER = 1,
+		APPEND_ADDINZIP = 2,
+	};
+
+	Error open(String path, ZipAppend append);
+	Error close();
+
+	Error start_file(String path);
+	Error write_file(Vector<uint8_t> data);
```
Comment (akien-mga) [inline modules/zip/zip_packer.h:59]: Use `p_` prefix.

**Review by akien-mga — APPROVED** [review]
Discussed in PR review meeting, we agreed to keep it as a module so it can be disabled if someone wants to.

Would be nice to fill the documentation but that could be done in a follow-up PR.

**Comment by MisterMX** [comment]
@akien-mga, I renamed the parameters and added the documentation.

**Review by fire — APPROVED** [review]
I skimmed over the code and didn't see anything obvious.

**Comment by akien-mga** [comment]
Thanks!

**Comment by Anutrix** [comment]
Can we cherry-pick this for 3.x if possible? It's been requested for 3.x before(https://github.com/godotengine/godot/pull/58900#issuecomment-1100205919).

**Comment by Calinou** [comment]
Documentation for some constants is missing in the class reference:

https://github.com/godotengine/godot/blob/69e5ad7f098730207249218236486e7ffa72f17f/modules/zip/doc_classes/ZIPPacker.xml#L65-L70

@MisterMX @jameswestman Could you provide a description for them? Thanks in advance :slightly_smiling_face:

**Comment by Macksaur** [comment]
I added documentation for these enum constants in https://github.com/godotengine/godot/pull/69677 as well as fixing a buffer-overrun and memory leak. (It's possible to use a specially crafted zip file to dump some stack contents). I also improved the support for longnames.

Can anyone take a look? Thanks.

**Pull request merged**